sgv2-dynamo-api/src/main/java/io/stargate/sgv2/dynamosvc/parser/FilterExpressionVisitor.java

```diff
@@ -0,0 +1,148 @@
+package io.stargate.sgv2.dynamosvc.parser;
+
+import com.amazonaws.services.dynamodbv2.model.AttributeValue;
+import com.google.common.base.Preconditions;
+import io.stargate.sgv2.dynamosvc.dynamo.DataMapper;
+import java.util.Map;
+import java.util.Objects;
+import org.antlr.v4.runtime.CharStream;
+import org.antlr.v4.runtime.CharStreams;
+import org.antlr.v4.runtime.CommonTokenStream;
+import org.antlr.v4.runtime.Lexer;
+import org.antlr.v4.runtime.tree.ParseTree;
+
+public class FilterExpressionVisitor extends ExpressionBaseVisitor<Object> {
+  private Map<String, String> nameMap;
+  private Map<String, AttributeValue> compareValueMap;
+  private Map<String, AttributeValue> item;
+
+  public FilterExpressionVisitor(
+      Map<String, String> nameMap, Map<String, AttributeValue> compareValueMap) {
+    Objects.requireNonNull(nameMap);
+    Objects.requireNonNull(compareValueMap);
+    this.nameMap = nameMap;
+    this.compareValueMap = compareValueMap;
+  }
+
+  /**
+   * This visitor walks through all conditions, and returns True if the given item satisfies the
+   * conditions and False otherwise.
+   */
+  public synchronized boolean isMatch(ParseTree tree, Map<String, AttributeValue> item) {
```

Comment:
Is this required? Looks like instances are created for every separate request and not (re)used concurrently?

Reply:
`synchronized` is indeed not required. I put this keyword thinking that the object could be (re)used by different threads, but it did not happen. I will remove this keyword.

The `item` is passed through this method so that this object could be reused across different `item`s. One request could lead to a number of `item`s as results, and I thought we could reduce some overhead by reusing the same object across `item`s for the same request.

Reply:
Right, that makes sense. Was thinking it might be something like this.